L63-python-eventlet-support.md

```diff
@@ -0,0 +1,108 @@
+because it will be blocked on the poll operation.
+
+This means that no Python application that uses eventlets can use gRPC.  This
+includes but is not limited to:
+
+- Using the etcd3 python library with eventlets
+
+- [OpenStack's Cinder](https://docs.openstack.org/cinder/latest/) using TooZ
+  with the etcd3 driver as a DLM
+
+- Using [cinderlib](https://docs.openstack.org/cinderlib/latest/) together with
+  gRPC to:
+
+  - An efficient imlementation of a
+    [CSI](https://github.com/container-storage-interface/spec) plugin such as
+    [Ember-CSI](https://ember-csi.io)
+
+  - oVirt using a gRPC service for their [Managed Block Storage feature](
+    https://www.ovirt.org/develop/release-management/features/storage/cinderlib-integration.html)
+
+  - [Swift object storage switched to plain HTTP from gRPC](
+    https://opendev.org/openstack/swift/commit/21d710b1f69b0e8e4e816d4c94c3ebb804456fd5?lang=sv-SE)
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+This gRFC proposes that we add an experimental custom I/O manager like the one
+we currently have for gevent.
+
+Eventlet and gevent use greenthreads differently so their implementation will
+not be exactly the same.
+
+The implementation will require:
+
+- Schedule asynchronous calls on greenlets so they can do the callbacks when
+  they complete their execution.
+
+- Make sure we schedule asynchronous calls on the right eventlet hub, the one
+  from the main thread where we do the polling.
+
+- A custom Event, because eventlet's cannot work between different native
+  threads, and we will get kicks from the main thread as well as the one with
+  the ThreadPoolExecutor.
+
+- Use a socket to wake the main thread hub so we don't have to wait until it
+  timeouts.
+
+- Signal the greenthread that is waiting on the accept socket call when we
+  close the socket so the server can gracefully stop.
+
+## Rationale
+
+### Alternative Design: Native threads
+
+The blocking of the polling can resolved if we use native threads for the
+polling methods that would block.
+
+This can be done using eventlet's `tpool.Proxy` class on the server and
+`completion_queue` and then passing removing the proxy when calling C code.
+
+This is how [Ember-CSI is currently solving the issue](
+https://github.com/embercsi/ember-csi/blob/148cc4a09db22df8aa4772dfb3696160cf37fe28/ember_csi/workarounds.py#L25-L54).
+
+The advantage of this approach is that the amount of code is small, but the
+downside is greater:
+
+- Server won't stop gracefully: Eventlet does not signal socket readers/writers
+  when the socket is closed from another greenthread, so the accept call that
+  will be awaiting won't know that the socket has been closed and will never
+  reach the point where it makes the accept callback with an error, thus
+  blocking the graceful stop of the gRPC server.
+
+- Performance is not good.  In tests the I/O manager has shown up to 4x better
+  performance.
```

Comment:
Alternatively, I see you are using `pipe` to synchronize in your implementation. You can let the poller thread not holding the GIL at all by implementing it in Cython or cpp. It should solve the performance problem. `asyncio` is moving toward this direction [2].

About the graceful shutdown, I don't have full context here. Is this an issue with cross-thread synchronization or cross-greenthread synchronization?

[2] https://github.com/grpc/grpc/pull/22259

Reply:
Thanks for the tip!
But even if I resolve the performance problem in that project it would still be a hack.  One that could be easily broken by changes in the gRPC library, and one that doesn't automatically benefit others, as they would have to find that project and then copy/paste that code in their own projects.

The graceful shutdown problem is caused by eventlet not signaling readers/writers when a socket is closed.  I added some comments in the implementation to explain this quirk: 
https://github.com/grpc/grpc/blob/3023219df9ef575e77cd190c2223f38eb301c974/src/python/grpcio/grpc/_cython/_cygrpc/grpc_eventlet.pyx.pxi#L70-L81

Reply:
About the graceful shutdown, if the eventlet application can inject exceptions from another greenthread to simulate socket close event, maybe it is also possible with a Python thread. For example, if the Python thread schedule a callback onto the eventlet thread, then the callback can perform the injection of exception.

Reply:
Yes that is exactly what the gRPC PR does, but doing it in the Ember-CSI workaround would be just another hack, requiring additional monkey patching.

It is true that now that I understand the gRPC code flow a little bit better I could probably improve the performance of the Ember-CSI hack, but it would still be that, a hack, and would not be easily used by others.